### PR TITLE
Enable rewrite_response to modify status and headers

### PR DESCRIPTION
**Pull request by ryanlovett**
This builds on @maresb 's work in #209 and enables one to also alter response HTTP headers and status code/reason.  Thanks to @yuvipanda for the idea!

### CODE REVIEW

File: tests/resources/jupyter_server_config.py [hunk 1]
```diff
@@ -40,10 +39,14 @@ def mappathf(path):
     'python-gzipserver': {
         'command': ['python3', './tests/resources/gzipserver.py', '{port}'],
     },
+    'python-http-rewrite-response': {
```
Comment (yuvipanda) [inline tests/resources/jupyter_server_config.py:42]: can you add tests here for code, headers, etc as well?

**Review by yuvipanda — APPROVED** [review]
Let's add a test, and then we're good to go!

File: jupyter_server_proxy/handlers.py [hunk 2]
```diff
@@ -46,7 +46,7 @@ def __init__(self, *args, **kwargs):
         self.host_allowlist = kwargs.pop('host_allowlist', ['localhost', '127.0.0.1'])
         self.rewrite_response = kwargs.pop(
             'rewrite_response',
-            lambda host, port, path, response: response.body
+            lambda request, host, port, path, response: {}
```
Comment (maresb) [inline jupyter_server_proxy/handlers.py:49]: It makes me really nervous to see a [mutable object inside a default argument](https://www.google.com/search?q=python+mutable+default+argument). But in this case I think it's effectively immutable thanks to the `lambda`.

**Comment by maresb** [comment]
I really like the extra functionality that this adds, thanks @ryanlovett!

I feel wary about this dictionary and the large number of function arguments flying around, and I wonder if we can do better...

What if we create a `RewritableResponse` object which is some sort of `dataclass` with properties `request`, `host`, `port`, `path`, `response`, `code`, `reason`, `headers`, and `body`. Then we can make the `rewrite_response` function have signature `RewritableResponse -> RewritableResponse`.

In addition to making the code simpler, this would also make it easy to compose rewrite functions or use them as decorators.

**Comment by yuvipanda** [comment]
While thinking through @maresb's point about unweily arguments, I realized what we should do is use https://github.com/jupyterhub/jupyter-server-proxy/blob/5fd3969efb31eae3b76a64860cf0de663ee55a92/jupyter_server_proxy/utils.py#L4. We use that [elsewhere](https://github.com/jupyterhub/jupyter-server-proxy/search?q=call_with_asked_args) to allow us to add arbitrary args in the future without breaking existing code. So users can request as few or as many args as they want, and we don't need to add extra layers of indirection (like an additional class).

**Comment by maresb** [comment]
Hmm, I'm quite fond of the ability to chain rewrites in my implementation. Do you see any way to achieve this with `call_with_asked_args`? It seems to me like it would be much harder that way, but maybe I just haven't thought enough in that direction.

**Comment by yuvipanda** [comment]
@maresb can you tell me a bit more about what you mean by 'chained rewrites'? An example would be lovely.

**Comment by maresb** [comment]
@yuvipanda, answered in #301.

**Pull request merged**